neps/nep-0413.md

```diff
@@ -0,0 +1,135 @@
+4) The signature should be simple to produce/verify, and transmitted securely.
+
+### Why Off Chain?
+So the user would not incur in GAS fees, nor the signed authentication gets broadcasted into a public network.
+
+### Why The Message Being Signed MUST NOT be a Transaction?
+An attacker could make the user inadvertently sign a valid transaction which, once signed, could be submitted into the network to execute it.
+
+### Why The Message Needs to Include a Service Identifier?
+To avoid replay attacks, i.e. a malicious app requesting the user to authenticate in their service, only to use the authentication message in a third-app. 
+
+Including the service's identifier (e.g. domain name) and making sure the user knows about it should mitigate this kind of attacks.
+
+### Why a FullAccess Key? Why Not Simply Creating an [FunctionCall Key](https://docs.near.org/concepts/basics/accounts/access-keys) for Signing?
+The most common flow for [NEAR user authentication into a Web3 frontend](https://docs.near.org/develop/integrate/frontend#user-sign-in--sign-out) involves the creation of a [FunctionCall Key](](https://docs.near.org/concepts/basics/accounts/access-keys)).
+
+One might feel tempted to reproduce such process here, for example, by creating a key that can only be used to call a non-existing method in the user's account. This is a bad idea for several reasons:
+1. It implies that the user needs to expend gas in creating a new key.
+2. Since any third-party can ask the user to create a `FunctionCall Key`, this opens a vector of attack.
+
+Using a FullAccess key allows us to be sure that the message was signed by the user (since nobody should have access to their `FullAccess Key`), while keeping the constraints of not expending gas in the process (because no new key needs to be created).
+
+### How to Return the Signed Message in a Safe Way
+Sending the signature in a query string to an arbitrary URL (even within the correct domain) is not secure as the data can be leaked (e.g. through headers, etc). Using URL fragments instead will improve security, since the fragment isn't included in the Referrer.
+
+### NEAR Signatures
+NEAR transaction signatures are not plain Ed25519 signatures but Ed25519 signatures of a SHA-256 hash (see [near/nearcore#2835](https://github.com/near/nearcore/issues/2835)). Any protocol that signs anything with NEAR account keys should use the same signature format.
+
+## Specification
+Given the previous [rationales](#rationale-and-alternatives), this NEP proposes the wallets to implement a `verifyOwner` method following the specifications bellow.
+
+### Goal
+The `verifyOwner` method must take as input a `domain` and a `message`, and transform them into a signature.
+
+### Input Interface
+In order to sign a message, `verifyOwner` must implement the following input interface:
+
+```jsx
+interface VerifyOwnerParams {
+  domain: string; // The app in which the user wants to authenticate, e.g. myapp.com.
```

Comment:
Is this the same message that is being sent by the entity requesting verification, or is the context added somewhere between the requester and the responder, i.e. by the wallet selector.

I think the requester shouldn't send this data because a bad implementation on the wallets side may use it to generate their hash rather than a domain they've actually verified.

Reply:
Sounds good!